File: generated_code/GeminiPro/Java/097_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructor for a Node.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Core Splay Operations ---

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. Brings the node with the given key (or the last
     * accessed node if the key is not found) to the root of the tree.
     * @param node The root of the tree (or subtree) to splay.
     * @param key The key to search for and bring to the root.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key is in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done
            if (node.left == null) {
                return node;
            }
            // Zig-Zig (Left Left)
            if (key < node.left.key) {
                node.left.left = splay(node.left.left, key);
                node = rightRotate(node);
            }
            // Zig-Zag (Left Right)
            else if (key > node.left.key) {
                node.left.right = splay(node.left.right, key);
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Do second rotation for Zig-Zig or Zig-Zag (or Zig)
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key is in the right subtree
        else {
            // Key is not in the tree, we are done
            if (node.right == null) {
                return node;
            }
            // Zig-Zag (Right Left)
            if (key < node.right.key) {
                node.right.left = splay(node.right.left, key);
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > node.right.key) {
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            }
            // Do second rotation for Zig-Zig or Zig-Zag (or Zig)
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    // --- Public API Methods ---

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root.
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree. If the key already exists,
     * the tree is splayed on that key but no new node is inserted.
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree on the key. This brings the closest node to the root.
        root = splay(root, key);

        // If key is already present, do nothing
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // If the new key is smaller than the root's key
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        }
        // If the new key is larger than the root's key
        else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree on the key. If the key exists, it becomes the root.
        root = splay(root, key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its largest element to its root.
            // Since all elements in the left subtree are smaller than the deleted key,
            // splaying for the deleted key will bring the max element to the top.
            leftSubtree = splay(leftSubtree, key);

            // The new root of the left subtree has no right child.
            // Attach the original right subtree there.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/097_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a dictionary-like set for integers
    using a Splay Tree.

    A splay tree is a self-adjusting binary search tree with the additional
    property that recently accessed elements are quick to access again.
    It performs basic operations such as insertion, look-up, and removal
    in O(log n) amortized time.

    This class provides the following methods:
    - insert(key): Adds an integer key to the set.
    - delete(key): Removes an integer key from the set.
    - search(key): Checks if a key exists and splays the accessed node.
    """

    class _Node:
        """A private node class for the Splay Tree."""
        def __init__(self, key, parent=None, left=None, right=None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty SplayTree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Splays the given node to the root of the tree through a series of rotations.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent

            if not grandparent:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        This method performs the splaying operation on the accessed node to move
        it to the root. If the key is not found, the last accessed node (the
        would-be parent) is splayed to the root instead.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        found = False

        while current:
            last_visited = current
            if key == current.key:
                found = True
                break
            elif key < current.key:
                current = current.left
            else:
                current = current.right

        if last_visited:
            self._splay(last_visited)

        return found

    def insert(self, key):
        """
        Inserts a key into the tree. If the key already exists, the tree
        is splayed but not modified.

        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the tree around the key. This brings the closest node to the root.
        self.search(key)

        # After splaying, if the key is already in the tree, it's at the root.
        if self.root.key == key:
            return  # Key already exists

        # Key does not exist, so insert the new node at the root and restructure.
        new_node = self._Node(key)

        if key < self.root.key:
            # New node becomes root; old root becomes its right child.
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.left = None
        else:  # key > self.root.key
            # New node becomes root; old root becomes its left child.
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.right = None

        self.root.parent = new_node
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree. If the key is not found, the tree is
        splayed around the closest key but is otherwise not modified.

        Args:
            key (int): The integer key to delete.
        """
        # Splay the tree. If the key exists, it becomes the new root.
        if not self.search(key) or self.root.key != key:
            # Key was not found, so nothing to delete. search() already splayed.
            return

        # At this point, the node to delete is the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree to be the new root.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # Promote the left subtree to be the new root.
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Both subtrees exist. We need to join them.
            # 1. Detach the left subtree.
            left_subtree.parent = None

            # 2. Find the maximum node in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # 3. Splay this max_node to the root of the (temporary) left subtree.
            #    We achieve this by temporarily setting the root for the splay operation.
            self.root = left_subtree
            self._splay(max_node)  # After this, self.root is max_node.

            # 4. Attach the original right subtree as the right child of the new root.
            #    The new root (max_node) is guaranteed to have no right child.
            self.root.right = right_subtree
            right_subtree.parent = self.root